clientconn_state_transition_test.go

```diff
@@ -0,0 +1,315 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package grpc
+
+import (
+	"net"
+	"sync"
+	"testing"
+	"time"
+
+	"golang.org/x/net/context"
+	"golang.org/x/net/http2"
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/internal/leakcheck"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/resolver/manual"
+)
+
+var testBalancer = &stateRecordingBalancer{}
+
+func init() {
+	balancer.Register(&stateRecordingBalancerBuilder{name: "state_recording_balancer", b: testBalancer})
+}
+
+func TestStateTransitions_SingleAddress(t *testing.T) {
+	for i, test := range []struct {
+		want   []connectivity.State
+		server func(net.Listener)
+	}{
+		// When the server returns server preface, the client enters READY.
+		{
+			want: []connectivity.State{
+				connectivity.Connecting,
+				connectivity.Ready,
+			},
+			server: func(lis net.Listener) {
+				conn, err := lis.Accept()
+				if err != nil {
+					t.Error(err)
+					return
+				}
+
+				framer := http2.NewFramer(conn, conn)
+				if err := framer.WriteSettings(http2.Setting{}); err != nil {
+					t.Errorf("Error while writing settings frame. %v", err)
+					return
+				}
+			},
+		},
+		// When the connection is closed, the client enters TRANSIENT FAILURE.
+		{
+			want: []connectivity.State{
+				connectivity.Connecting,
+				connectivity.TransientFailure,
+			},
+			server: func(lis net.Listener) {
+				conn, err := lis.Accept()
+				if err != nil {
+					t.Error(err)
+					return
+				}
+
+				conn.Close()
+			},
+		},
+		// When a READY connection is closed, the client enters TRANSIENT FAILURE before CONNECTING.
+		{
+			want: []connectivity.State{
+				connectivity.Connecting,
+				connectivity.Ready,
+				connectivity.TransientFailure,
+				connectivity.Connecting,
+			},
+			server: func(lis net.Listener) {
+				conn, err := lis.Accept()
+				if err != nil {
+					t.Error(err)
+					return
+				}
+
+				framer := http2.NewFramer(conn, conn)
+				if err := framer.WriteSettings(http2.Setting{}); err != nil {
+					t.Errorf("Error while writing settings frame. %v", err)
+					return
+				}
+
+				// Prevents race between onPrefaceReceipt and onClose.
+				time.Sleep(50 * time.Millisecond)
```

Comment:
This is unfortunate and probably needs to be changed. I thought about having the server also watch the notifications in `testBalancer` until it sees READY before doing this, but:

- You'd have to set up a goroutine ahead of sending the preface receipt to avoid a race between the server starting to watch and the client changing state to READY.
- The state change is currently just a single channel that's watched by the test on L147. Adding another watcher - the server - might means you'd have to have some way to send each state change to multiple watchers, which can get complicated.

... and both of those could get a bit ugly. WDYT @dfawley ?

Reply:
Fixed by making the test artisanal and causing it to wait for READY to be seen.